File: app/src/main/java/com/chris/eban/presenter/event/EventDetailActivity.java
package com.chris.eban.presenter.event;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.view.Window;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import com.chris.eban.R;
import com.chris.eban.databinding.ActivityEventDetailBinding;
import com.chris.eban.presenter.BaseActivity;

import java.util.Objects;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import timber.log.Timber;

public class EventDetailActivity extends BaseActivity {

    public static final String PAGE_STATUS = "com.chris.eban.presenter.event.EventDetailActivity.pageStatus";
    public static final String PAGE_ITEM = "com.chris.eban.presenter.event.EventDetailActivity.pageItem";
    public static final String PAGE_STATUS_EDIT = "edit"; // 编辑状态
    public static final String PAGE_STATUS_SAVE = "save"; // 保存状态
    private static final String TAG = "EventDetailActivity";

    private String status;
    @Inject
    ViewModelProvider.Factory factory;

    private boolean saved = false;

    private ActivityEventDetailBinding binding;
    private EventDetailViewModel viewModel;
    private TextWatcher watcher = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @Override
        public void afterTextChanged(Editable s) {

        }
    };

    private InputMethodManager imm;

    private View.OnTouchListener touchListener = new View.OnTouchListener() {
        @SuppressLint("ClickableViewAccessibility")
        @Override
        public boolean onTouch(View v, MotionEvent event) {
            showIme((EditText) v);
            return true;
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_event_detail);
        binding.setLifecycleOwner(this);
        setSupportActionBar(binding.toolbar2);

        setStatusBarTextColor(getWindow(), true);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowTitleEnabled(false);
        }

        viewModel = ViewModelProviders.of(this, factory).get(EventDetailViewModel.class);
        // 判断启动方式，切换编辑状态
        Intent intent = getIntent();
        status = intent.getStringExtra(PAGE_STATUS);
        Timber.tag(TAG).d("page start status: %s", status);
        EventItem item = Objects.requireNonNull(intent.getExtras()).getParcelable(PAGE_ITEM);
        viewModel.setItem(item);
        binding.setDetail(viewModel);

        binding.etTitle.setOnTouchListener(touchListener);
        binding.etTitle.addTextChangedListener(watcher);
        binding.etContent.setOnTouchListener(touchListener);
        binding.etContent.addTextChangedListener(watcher);
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        switch (status) {
            case PAGE_STATUS_EDIT:
                showToolBarMenu(menu, true);
                break;
            case PAGE_STATUS_SAVE:
                showToolBarMenu(menu, false);
                break;
        }
        return super.onPrepareOptionsMenu(menu);
    }

    private void showToolBarMenu(Menu menu, boolean toggle) {

        menu.findItem(R.id.action_event_save).setVisible(toggle);
        menu.findItem(R.id.action_event_edit).setVisible(!toggle);
        menu.findItem(R.id.action_event_favorite).setVisible(!toggle);
        menu.findItem(R.id.action_event_share).setVisible(!toggle);
        menu.findItem(R.id.action_event_delete).setVisible(!toggle);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.event_save, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_event_save:
                saveEvent();
                return true;
            case R.id.action_event_edit:
                showIme();
                break;
            case android.R.id.home:
                hideIme();
                finish();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void hideIme() {
        if (imm == null)
            imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(binding.etTitle.getWindowToken(), 0);
        binding.etContent.clearFocus();
        binding.etTitle.clearFocus();
        status = PAGE_STATUS_SAVE;
        invalidateOptionsMenu();
    }

    private void showIme() {
        showIme(binding.etTitle);
    }

    @SuppressLint("ClickableViewAccessibility")
    private void showIme(EditText view) {
        if (imm == null)
            imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
        if (imm != null) {
            view.requestFocus();
            view.setSelection(view.getText().length());
            imm.showSoftInput(view, InputMethodManager.RESULT_SHOWN);
            status = PAGE_STATUS_EDIT;
            invalidateOptionsMenu();
            view.setOnTouchListener(null);
        }
    }


    public void setStatusBarTextColor(Window window, boolean lightStatusBar) {
        // 设置状态栏字体颜色 白色与深色
        View decor = window.getDecorView();
        int ui = decor.getSystemUiVisibility();
        ui |= View.SYSTEM_UI_FLAG_LAYOUT_STABLE;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (lightStatusBar) {
                ui |= View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
            } else {
                ui &= ~View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
            }
        }
        decor.setSystemUiVisibility(ui);
    }

    private void saveEvent() {

        if (saved) return;
        saved = true;
        hideIme();

        Editable title = binding.etTitle.getText();
        Editable content = binding.etContent.getText();

        if (TextUtils.isEmpty(title) && TextUtils.isEmpty(content)) {
            setResult(RESULT_CANCELED);
            finish();
        } else {
            Timber.tag(TAG).d("\nEventTitle:%s \nEventContent:%s", title, content);
            viewModel.setItem(title, content);
            viewModel.saveEvent();
            setResult(RESULT_OK);
        }
    }
}

File: app/src/main/java/com/chris/eban/presenter/event/EventDetailActivity.kt
package com.chris.eban.presenter.event

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.Window
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.chris.eban.R
import com.chris.eban.databinding.ActivityEventDetailBinding
import com.chris.eban.presenter.BaseActivity
import timber.log.Timber
import javax.inject.Inject

class EventDetailActivity : BaseActivity() {

    private var status: String? = null

    @Inject
    lateinit var factory: ViewModelProvider.Factory

    private var binding: ActivityEventDetailBinding? = null
    private var viewModel: EventDetailViewModel? = null

    private var imm: InputMethodManager? = null

    private val touchListener = View.OnTouchListener { v, _ ->
        showIme(v as EditText)
        true
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_event_detail)
        binding!!.lifecycleOwner = this

        setSupportActionBar(binding!!.toolbar2)

        setStatusBarTextColor(window, true)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowTitleEnabled(false)

        viewModel = ViewModelProvider(this, factory).get(EventDetailViewModel::class.java)
        // 判断启动方式，切换编辑状态
        val intent = intent
        status = intent.getStringExtra(PAGE_STATUS)
        Timber.tag(TAG).d("page start status: %s", status)

        val itemId = intent.getLongExtra(PAGE_ITEM_ID, -1L)
        if (itemId != -1L)
            viewModel?.queryItemById(itemId)

        binding!!.detail = viewModel

        binding!!.etTitle.setOnTouchListener(touchListener)
        binding!!.etContent.setOnTouchListener(touchListener)
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        when (status) {
            PAGE_STATUS_EDIT -> showToolBarMenu(menu, true)
            PAGE_STATUS_SAVE -> showToolBarMenu(menu, false)
        }
        return super.onPrepareOptionsMenu(menu)
    }

    private fun showToolBarMenu(menu: Menu, toggle: Boolean) {

        menu.findItem(R.id.action_event_save).isVisible = toggle
        menu.findItem(R.id.action_event_edit).isVisible = !toggle
        menu.findItem(R.id.action_event_favorite).isVisible = !toggle
        menu.findItem(R.id.action_event_share).isVisible = !toggle
        menu.findItem(R.id.action_event_delete).isVisible = !toggle
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.event_save, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_event_save -> {
                saveEvent()
                return true
            }
            R.id.action_event_edit -> showIme()
            android.R.id.home -> {
                hideIme()
                finish()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun hideIme() {
        if (imm == null)
            imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm!!.hideSoftInputFromWindow(binding!!.etTitle.windowToken, 0)
        binding!!.etContent.clearFocus()
        binding!!.etTitle.clearFocus()
        binding!!.etTitle.setOnTouchListener(touchListener)
        binding!!.etContent.setOnTouchListener(touchListener)
        status = PAGE_STATUS_SAVE
        invalidateOptionsMenu()
    }

    private fun showIme() {
        showIme(binding!!.etContent)
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun showIme(view: EditText) {
        if (imm == null)
            imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        if (imm != null) {
            view.requestFocus()
            view.setSelection(view.text.length)
            imm!!.showSoftInput(view, InputMethodManager.RESULT_SHOWN)
            status = PAGE_STATUS_EDIT
            invalidateOptionsMenu()
            view.setOnTouchListener(null)
        }
    }


    private fun setStatusBarTextColor(window: Window, lightStatusBar: Boolean) {
        // 设置状态栏字体颜色 白色与深色
        val decor = window.decorView
        var ui = decor.systemUiVisibility
        ui = ui or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            ui = if (lightStatusBar) {
                ui or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
            } else {
                ui and View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR.inv()
            }
        }
        decor.systemUiVisibility = ui
    }

    private fun saveEvent() {

        hideIme()

        val title = binding!!.etTitle.text
        val content = binding!!.etContent.text

        if (TextUtils.isEmpty(title) && TextUtils.isEmpty(content)) {
            if (viewModel!!.hadId()) viewModel!!.deleteEvent()
            setResult(Activity.RESULT_CANCELED)
            finish()
        } else {
            Timber.tag(TAG).d("\nEventTitle:%s \nEventContent:%s", title, content)
            viewModel!!.setItem(title, content)
            viewModel!!.saveEvent()
            setResult(Activity.RESULT_OK)
        }
    }

    companion object {
        const val PAGE_STATUS = "com.chris.eban.presenter.event.EventDetailActivity.pageStatus"

        const val PAGE_ITEM_ID = "com.chris.eban.presenter.event.EventDetailActivity.pageItemId"

        const val PAGE_STATUS_EDIT = "edit" // 编辑状态
        const val PAGE_STATUS_SAVE = "save" // 保存状态
        private const val TAG = "EventDetailActivity"
    }
}
